sympy/matrices/eigen.py

```diff
@@ -19,6 +22,62 @@
 from .utilities import _iszero
 
 
+def _eigenvals_triangular(M, multiple=False):
+    """A fast decision for eigenvalues of an upper or a lower triangular
+    matrix.
+    """
+    diagonal_entries = [M[i, i] for i in range(M.rows)]
+    if multiple:
+        return diagonal_entries
+    return dict(Counter(diagonal_entries))
+
+
+def _eigenvals_eigenvects_mpmath(M):
+    prec = max([x._prec for x in M.atoms(Float)])
+
+    A = mp.matrix(M.evalf(n=prec_to_dps(prec)))
```

Comment:
oops...we are always using the same `prec`

Reply:
I think that mpmath already gives different precision result for eigenvalues even if the precision for matrix elements are specified.

```python3
import mpmath as mp

a = mp.mpf('1.123', prec=80)
b = mp.mpf('3.43', prec=90)
c = mp.mpf('4.57', prec=100)
d = mp.mpf(6)

m = mp.matrix([[a, b], [c, d]])
E, EV = mp.eig(m)

for x in E:
    print(x._mpf_)
    
for x in EV:
    print(x._mpf_)
```
```
(1, 306351681749411, -48, 49)
(0, 1155648970429707, -47, 51)
(0, 7570250651444985, -53, 53)
(0, 3952680198505023, -53, 52)
(1, 4880670393388953, -53, 53)
(0, 8168541841269819, -53, 53)
```

Reply:
So it looks like mpmath will only return precision of 15 digits regardless of what is passed...if so then what you had before is better -- just pass `matrix.n()` to mpmath.

Reply:
It is possible to change the precision: https://github.com/fredrik-johansson/mpmath/blob/ffec62b29b6fd4a3a5e8b18f6a35896de2f51103/doc/source/basics.txt#L70

Reply:
Very helpful pointer, thanks! So the question I have is what precision to select when there is potentially a range of precisions present in a matrix. And the matter is complicated when there are functions present, too. I suppose the safest thing to do is just use the most precise value but realize that the actual precision could be significantly less.

Reply:
I am not sure what precision to use for the calculation but I suspect that it should be calculated by seeing how the deltas of each component of E change:
```python
def _eigenvals_eigenvects_mpmath(M):
    from mpmath import workprec
    norm2 = lambda v1, v2: mp.fsum([(r1 - r2)**2 + (i1 - i2)**2
        for (r1,i1), (r2,i2) in zip(v1, v2)])
    v = lambda E: [(mp.re(e), mp.im(e)) for e in E]

    norm = None
    new_prec = max([x._prec for x in M.atoms(Float)])
    while new_prec < DEFAULT_MAXPREC:

        with workprec(new_prec):
            A = mp.matrix(M.evalf(n=prec_to_dps(new_prec)))
            E, ER = mp.eig(A)

        if norm is not None:
            v1 = v2
            v2 = v(E)
            norm_new = norm2(v1, v2)
            if abs(norm - norm_new) <  50*2**-prec:
                return E, ER
        else:
            v2 = v(E)
            norm_new = mp.fsum([r**2 + i**2 for r,i in v2])

        prec = new_prec
        new_prec *= 2
        norm = norm_new

    # we get here because the next step would have taken us
    # past MAXPREC or because we never took a step; in case
    # of the latter, we refuse to send back a solution since
    # it would not have been verified; we also resist taking
    # a small step to arrive exactly at MAXPREC since then
    # the norm and new_norm would be artificially close
    raise PrecisionExhausted
```